### PR TITLE
[MOKA-001][MOKA-002][MOKA-003]

### PR DESCRIPTION
**Pull request by titopandub**
Fixing some issue from comments.

### CODE REVIEW

File: app/controllers/items_controller.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
 class ItemsController < ApplicationController
+  before_filter :set_item, only: [:show, :edit, :update, :destroy]
```
Comment (glaksmono) [inline app/controllers/items_controller.rb:2]: Deprecated: https://github.com/rails/rails/blob/v5.0.0.beta2/actionpack/lib/abstract_controller/callbacks.rb#L190-L193

File: app/controllers/items_controller.rb [hunk 2]
```diff
@@ -10,15 +11,44 @@ def index
...
 
+  def new
+    @item = current_business.items.new
+  end
+
   def create
-    # TODO: [MOKA-001] Finish the implementation of Create/Update/Delete for Items
+    @item = current_business.items.new(item_params)
+    if @item.save
+      redirect_to @item, notice: 'item was successfully created.'
```
Comment (glaksmono) [inline app/controllers/items_controller.rb:21]: Could have been using locale docs: http://guides.rubyonrails.org/i18n.html

i.e. using `t(...)`

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,11 +1,47 @@
...
+    else
+      render :edit
+    end
+  end
+
+  private
+  def set_user
+    @user = User.find(params[:id])
+    if @user.id != current_user.id
+      redirect_to root_path and return false
```
Comment (glaksmono) [inline app/controllers/users_controller.rb:38]: what is this and what happened with the session?

File: app/views/layouts/_header.html.erb [hunk 4]
```diff
@@ -4,7 +4,9 @@
     <ul class="nav navbar-nav navbar-right">
       <% if current_user %>
         <ul>
-          <li><%= link_to "Create Business", new_business_path %></li>
+          <% if current_user.business.blank? || current_user.business.new_record? %>
```
Comment (glaksmono) [inline app/views/layouts/_header.html.erb:7]: why using `.new_record` at all? http://apidock.com/rails/ActiveRecord/Base/new_record%3F

why not making sure that every time business is created, it is saved..

File: test/controllers/businesses_controller_test.rb [hunk 5]
```diff
@@ -1,7 +1,41 @@
...
+  setup do
+    @user = User.create!(email: "test@test.com", password: "testpassword",
+                         first_name: "Test", last_name: "Test")
+    @business_hash = {
+      name: "Test Business", address: "Test Address",
+      city: "Jakarta", user: @user
+    }
+  end
+
+  test 'should get new' do
```
Comment (glaksmono) [inline test/controllers/businesses_controller_test.rb:18]: what is this test case about?

File: test/controllers/businesses_controller_test.rb [hunk 6]
```diff
@@ -1,7 +1,41 @@
...
+  end
+
+  test 'should create business' do
+    login(@user)
+    assert_difference('Business.count') do
+      post :create, business: @business_hash
+    end
+
+    assert_redirected_to items_path
+  end
```
Comment (glaksmono) [inline test/controllers/businesses_controller_test.rb:40]: no test case for update? how about the failed cases?

File: test/controllers/items_controller_test.rb [hunk 7]
```diff
@@ -1,7 +1,96 @@
...
+  end
+
+  def create_item
+    @item = Item.create!(name: "Green Tea Latte", price: 35_000, business: @business)
+  end
+
+  setup do
+    @user = User.create!(email: "test@test.com", password: "testpassword",
+                         first_name: "Test", last_name: "Test")
+    @item_hash = { name: "Double shot", price: 40_000 }
```
Comment (glaksmono) [inline test/controllers/items_controller_test.rb:16]: price is `40_000`? any reason why not using `bigint` or just `integer`?

File: test/controllers/home_controller_test.rb [hunk 8]
```diff
@@ -1,7 +1,20 @@
...
+    login(@user)
+
+    get :index
+    assert_redirected_to items_path
+  end
+
+  test 'when not login, should redirected to login' do
+    get :index
+    assert_redirected_to login_path
+  end
```
Comment (glaksmono) [inline test/controllers/home_controller_test.rb]: ah looks like it's here - https://github.com/glaksmono/moka-rails-interview/pull/4/files#diff-6d313d1fc37182e0aa18b6b426cf2619R21

File: test/controllers/users_controller_test.rb [hunk 9]
```diff
@@ -1,7 +1,97 @@
...
+    login(user)
+
+    patch :update, id: user, user: @user_hash
+    assert_redirected_to user_path(assigns(:user))
+    assert_equal 'testupdate@test.com', assigns(:user).email
+    assert_equal 'Test Update', assigns(:user).first_name
+    assert_equal 'Test Update', assigns(:user).last_name
+  end
+
+  test 'when editing other user profile, should redirected back to root' do
```
Comment (glaksmono) [inline test/controllers/users_controller_test.rb:89]: 👍

File: test/controllers/users_controller_test.rb [hunk 10]
```diff
@@ -1,7 +1,97 @@
...
+
+    assert_redirected_to items_path
+  end
+
+  test 'should fail create user on invalid params' do
+    @user_hash.delete(:password)
+    post :create, user: @user_hash
+
+    assert_template :new
+  end
```
Comment (glaksmono) [inline test/controllers/users_controller_test.rb:49]: how about password length? and different parameter that are gone?

File: test/test_helper.rb [hunk 11]
```diff
@@ -3,8 +3,13 @@
...
-  # Setup all fixtures in test/fixtures/*.yml for all tests in alphabetical order.
-  fixtures :all
+  def login(user)
+    session[:user_id] = user.id
+  end
 
-  # Add more helper methods to be used by all tests here...
+  teardown do
+    (ActiveRecord::Base.connection.tables - %w{schema_migrations}).each do |table_name|
+      ActiveRecord::Base.connection.execute "TRUNCATE TABLE #{table_name};"
```
Comment (glaksmono) [inline test/test_helper.rb:12]: why not just run `rake db:migrate VERSION=0`?

File: app/controllers/items_controller.rb (hunk 1 above)
Comment (titopandub) [reply app/controllers/items_controller.rb:2]: Your Gemfile is using Rails 4.2.5.1, so it's YAGNI to see Rails 5 right now, unless you want to upgrade.

File: app/controllers/items_controller.rb (hunk 2 above)
Comment (titopandub) [reply app/controllers/items_controller.rb:21]: YAGNI, there is no need to have i18n right now, you don't have language switcher yet.

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (titopandub) [reply app/controllers/users_controller.rb:38]: The session only securing unauthenticated user, but not limit authorisation between user. It prevent other user edit. Your routes is still using resources on user.

File: app/views/layouts/_header.html.erb (hunk 4 above)
Comment (titopandub) [reply app/views/layouts/_header.html.erb:7]: It's used for hiding create business link only when you have saved the record. When you are still in the form, then the link is still there.

File: test/controllers/businesses_controller_test.rb (hunk 5 above)
Comment (titopandub) [reply test/controllers/businesses_controller_test.rb:18]: It's for the form.

File: test/controllers/items_controller_test.rb (hunk 7 above)
Comment (titopandub) [reply test/controllers/items_controller_test.rb:16]: Sorry, I don't understand this.

File: test/controllers/users_controller_test.rb (hunk 10 above)
Comment (titopandub) [reply test/controllers/users_controller_test.rb:49]: You should test that on model.

File: test/test_helper.rb (hunk 11 above)
Comment (titopandub) [reply test/test_helper.rb:12]: It's for cleaning the DB.

**Pull request closed without merging**